File: backend/app/api/contacts..py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.database.dependencies import get_db
from app.models.user import User
from app.schemas.contact import (
    ContactCreate,
    ContactResponse,
    ContactUpdate,
)
from app.services.contact_service import (
    create_contact,
    delete_contact,
    get_contact,
    get_contacts,
    update_contact,
)


router = APIRouter(
    prefix="/contacts",
    tags=["Contacts"],
)

@router.post(
    "",
    response_model=ContactResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_new_contact(
    data: ContactCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return create_contact(
        db=db,
        user_id=current_user.id,
        data=data,
    )

@router.get(
    "",
    response_model=list[ContactResponse],
)
def list_contacts(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_contacts(
        db=db,
        user_id=current_user.id,
    )

@router.get(
    "/{contact_id}",
    response_model=ContactResponse,
)
def get_single_contact(
    contact_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    contact = get_contact(
        db=db,
        user_id=current_user.id,
        contact_id=contact_id,
    )

    if contact is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contact not found.",
        )

    return contact

@router.patch(
    "/{contact_id}",
    response_model=ContactResponse,
)
def update_existing_contact(
    contact_id: int,
    data: ContactUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    contact = get_contact(
        db=db,
        user_id=current_user.id,
        contact_id=contact_id,
    )

    if contact is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contact not found.",
        )

    return update_contact(
        db=db,
        contact=contact,
        data=data,
    )

@router.delete(
    "/{contact_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_existing_contact(
    contact_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    contact = get_contact(
        db=db,
        user_id=current_user.id,
        contact_id=contact_id,
    )

    if contact is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contact not found.",
        )

    delete_contact(
        db=db,
        contact=contact,
    )

    return None

File: frontend/src/api/contacts.ts
import apiClient from "./client";

export interface Contact {
  id: number;
  user_id: number;
  name: string;
  email?: string | null;
  phone?: string | null;
  company?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateContactData {
  name: string;
  email?: string;
  phone?: string;
  company?: string;
  notes?: string;
}

export interface UpdateContactData {
  name?: string;
  email?: string;
  phone?: string;
  company?: string;
  notes?: string;
}

export async function createContact(
  data: CreateContactData
): Promise<Contact> {
  const response = await apiClient.post<Contact>(
    "/contacts",
    data
  );

  return response.data;
}

export async function getContacts(): Promise<Contact[]> {
  const response = await apiClient.get<Contact[]>(
    "/contacts"
  );

  return response.data;
}

export async function getContact(
  contactId: number
): Promise<Contact> {
  const response = await apiClient.get<Contact>(
    `/contacts/${contactId}`
  );

  return response.data;
}

export async function updateContact(
  contactId: number,
  data: UpdateContactData
): Promise<Contact> {
  const response = await apiClient.patch<Contact>(
    `/contacts/${contactId}`,
    data
  );

  return response.data;
}

export async function deleteContact(
  contactId: number
): Promise<void> {
  await apiClient.delete(`/contacts/${contactId}`);
}
